src/image/PhotoCalib.cc

```diff
@@ -94,6 +96,31 @@ void toNanojanskyVariance(ndarray::Array<float const, 2, 1> const &instFlux,
                (eigenInstFluxVar / eigenInstFlux.square() + (scaleErr / eigenFlux * eigenInstFlux).square());
 }
 
+/**
+ * Compute the variance of an array of fluxes, for calculations on nJy calibrated MaskedImages.
+ *
+ * MaskedImage stores the variance instead of the standard deviation, so we can skip a sqrt().
+ * Usage in calibrateImage() is to compute instFlux (ADU) directly, so that the calibration scale is
+ * implictly `flux/instFlux` (and thus not passed as an argument).
+ *
+ * @param flux[in] The flux in nJy.
+ * @param fluxVar[in] The variance of the fluxes.
+ * @param scaleErr[in] The error on the calibration scale.
+ * @param instFlux[in] The instrumental fluxes calculated from the fluxes.
+ * @param out[out] The output array to fill with the instrumental variance values.
+ */
+void fromNanojanskyVariance(ndarray::Array<float const, 2, 1> const &flux,
+                            ndarray::Array<float const, 2, 1> const &fluxVar, float scaleErr,
+                            ndarray::Array<float const, 2, 1> const &instFlux,
+                            ndarray::Array<float, 2, 1> out) {
+    auto eigenFlux = ndarray::asEigen<Eigen::ArrayXpr>(flux);
+    auto eigenFluxVar = ndarray::asEigen<Eigen::ArrayXpr>(fluxVar);
+    auto eigenInstFlux = ndarray::asEigen<Eigen::ArrayXpr>(instFlux);
+    auto eigenOut = ndarray::asEigen<Eigen::ArrayXpr>(out);
+    eigenOut = (eigenFluxVar / eigenFlux.square() - (scaleErr / eigenFlux * eigenInstFlux).square()) *
```

Comment:
```suggestion
    eigenOut = (eigenFluxVar / eigenFlux.square() - (scaleErr / (eigenFlux * eigenInstFlux).square())) *
```
I don't remember the C++ precedence rules well enough to know if what you had was already correct, but this ensures it.

Reply:
Uh oh: the above is the exact expression used in the `ADU->nJy` version:

```
    eigenOut = eigenFlux.square() *
               (eigenInstFluxVar / eigenInstFlux.square() + (scaleErr / eigenFlux * eigenInstFlux).square());
```

Have we been doing that incorrectly the whole time? Because your change here does switch the precedence. I've never been sure whether that calibration error is a variance or not. In jointcal just copied it over from PhotoCalTask: https://github.com/lsst/pipe_tasks/blob/main/python/lsst/pipe/tasks/photoCal.py#L418

Reply:
Oh, wait, I read that incorrectly.  Your original version was correct (but not obviously so), and I was trying to write it as
```
    eigenOut = (eigenFluxVar / eigenFlux.square() - (scaleErr / (eigenFlux / eigenInstFlux).square())) *
```
(note the divide rather than multiply inside the new parentheses).

So the formula you had looks correct (though maybe not obviously so), and we've got a strange test tolerance to explain.

Reply:
Oh, and I apparently put the closing parentheses in the wrong place in my original suggestion to (now fixed in the post above).  Here's a fully corrected suggestion:
```suggestion
    eigenOut = (eigenFluxVar / eigenFlux.square() - (scaleErr / (eigenFlux / eigenInstFlux)).square()) *
```

Reply:
Hmmm... I believe the original expression evaluates to `(scaleErr / eigenFlux) * eigenInstFlux`, because `/` and `*` have the same precedence, and are evaluated left-to-right, not the version that you wrote above.

Reply:
But unfortunately, changing to make that `flux/instFlux` division explicit didn't fix the numerical problem.

Reply:
The way I think about it - which may be worth writing out more explicitly via another temporary variable for `scale` - is
```
scale = flux / instFlux
instFluxVar = instFlux**2 * (fluxVar / flux**2) - (scaleVar / scale**2)
```

Reply:
Would it be possible to modify the numbers going into the test make sure we're not adding/subtracting a big number to/from a small number here?

Reply:
As we discussed on slack, I think we both now agree that the `PhotoCalib.calibrationErr` is supposed to be `sigma^2` (that's what comes out of `PhotoCalTask`), although my docstring is not really specific enough, sadly.

http://doxygen.lsst.codes/stack/doxygen/x_mainDoxyDoc/classlsst_1_1afw_1_1image_1_1_photo_calib.html#adb7c4cc78845e8d64666b3751cf5a51b